### PR TITLE
Implement indexing methods for sparse tensors

### LINKED ISSUE #7416: [feature request] `select_index` for sparse tensors
## Issue description

Sparse tensors do not currently support indexing (neither `gather`, nor `select_index`). This could be handy for e.g. embedding categorical features with non-contiguous vocabulary. 

Sidenote: a similar result can be achieved via bucketization (see #7284), without using sparse tensors.

## Code example

```
>>> import numpy as np
>>> import torch
>>> vocab = [102, 104, 2, 103, 0, 101, 3]
>>> lookup_table = torch.sparse.IntTensor(
...     torch.tensor(np.atleast_2d(vocab)), 
...     torch.arange(len(vocab), dtype=torch.int), 
...     torch.Size((max(vocab) + 1, )))
...
>>> lookup_table[102]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Sparse tensors do not have strides.
```

## System Info

```
PyTorch version: 0.5.0a0+3e785d5
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.3
GCC version: Could not collect
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.14.2)
[pip3] torch (0.5.0a0+3e785d5)
[pip3] torchvision (0.2.1)
[conda] torch                     0.4.0                     <pip>
[conda] torch                     0.5.0a0+3e785d5           <pip>
[conda] torchvision               0.2.1                     <pip>
```

### PR DESCRIPTION
**Pull request by pearu**
Resolves #7416 .

This PR implements the following indexing methods for sparse tensors:
-  [x] `select`
-  [x] `index_select`

Note that this PR also modifies [gen.py](https://github.com/pytorch/pytorch/pull/24937/files#diff-76aa8cb3d0fad99c5f761d08cbcb4d19) that is not directly required to resolve the original issue but to work around a CI build issue reported in issue #24931 .

### CODE REVIEW

**Comment by pearu** [comment]
gen.py changes can be undone after #25031 is merged.

**Review by zou3519 — COMMENTED** [review]
The sparse_select code looks pretty good. I had some initial minor comments / questions. Will come back to review the sparse index_select code soon.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -452,6 +452,38 @@ Tensor reshape_as(const Tensor& self, const Tensor& other) {
...
+
+  auto indices = self._indices();
+  auto values = self._values();
+  auto new_sizes = self.sizes().vec();
+  new_sizes.erase(new_sizes.begin() + dim);
+
+  if (dim < sparse_dim) {
+    auto nzIndices = (indices[dim] == index).nonzero().view(-1);
+    auto new_values = values.index_select(0, nzIndices);
+    if (sparse_dim == 1) {
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/TensorShape.cpp]: I agree that 0-D sparse tensor would be preferred. One could think of it as a scalar field of dense tensors.. Unfortunately, 0-D sparse tensors appear to be unsupported.

File: aten/src/ATen/gen.py [hunk 2]
```diff
@@ -369,6 +369,15 @@ def cmpfiles_with_eol_normalization(a, b, names):
                 results[0].append(x)
             else:
                 results[1].append(x)
+                import difflib
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/gen.py]: When this code is executed, it was really difficult to guess what was wrong, especially when this happened only within few CI instances, in most CI cases as well as locally, the code was never executed. So, showing the difference was helpful to continue debugging why the files were different.

File: aten/src/ATen/SparseTensorImpl.cpp [hunk 3]
```diff
@@ -55,7 +55,7 @@ bool SparseTensorImpl::is_contiguous(at::MemoryFormat memory_format) const {
   AT_ERROR("sparse tensors do not have is_contiguous");
 }
 int64_t SparseTensorImpl::stride(int64_t d) const {
-  AT_ERROR("sparse tensors do not have strides");
+  AT_ERROR("sparse tensors do not have stride");
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/SparseTensorImpl.cpp]: ok

File: aten/src/ATen/gen.py [hunk 4]
```diff
@@ -444,8 +453,9 @@ def generate_outputs():
         if len(mismatch) > 1:
             file_component = '{' + file_component + '}'
         update_cmd = "cp {}/{} {}".format(core_install_dir, file_component, core_source_path)
-        raise RuntimeError("Source files: {} did not match generated files.  To update the source files, "
-                           "set environment variable GEN_TO_SOURCE or run \"{}\"".format(mismatch, update_cmd))
+        import warnings
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/gen.py]: Making it an error will cause couple of CI cases to fail with no clear reason why. This seems to be related to https://github.com/pytorch/pytorch/issues/24931 .

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/gen.py]: I agree with Richard, this is a land blocker. Do you have the failing CI cases on hand?

**Review by ezyang — CHANGES_REQUESTED** [review]
The assert that generated and checked in files must be a hard error. If you are seeing mysterious CI failures please tell us about them.

File: aten/src/ATen/gen.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/gen.py]: I'll enable the RuntimeError in the next commit, this should get the CI cases.

File: aten/src/ATen/native/TensorShape.cpp [hunk 5]
```diff
@@ -452,6 +452,38 @@ Tensor reshape_as(const Tensor& self, const Tensor& other) {
   return self.reshape(other.sizes());
 }
 
+static Tensor select_sparse(const Tensor& self, int64_t dim, int64_t index) {
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/TensorShape.cpp]: yes, done.

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/TensorShape.cpp]: Note that `select` is also used in getitem methods. Erroring out if sparse dim is `1` would be unexpected.

File: aten/src/ATen/gen.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/gen.py]: See, for instance, at 20:23:57 of
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-build/47580/console

**Comment by ezyang** [comment]
Failure in ROCm is indeed mysterious:

```
20:23:35 --------------------------------------------------------------------------------
20:23:35 x=TensorMethods.h, a=/var/lib/jenkins/workspace/build/aten/src/ATen/core_tmp, b=/var/lib/jenkins/workspace/cmake/../aten/src/ATen/core
20:23:35  4203: -             break;
20:23:35  4204: -         case Backend::SparseCPU:
20:23:35  4205: -             return SparseCPUType::index_select(const_cast<Tensor&>(*this), dim, index);
20:23:35 --------------------------------------------------------------------------------
20:23:35 Traceback (most recent call last):
20:23:35   File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 465, in <module>
20:23:35     generate_outputs()
20:23:35   File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 457, in generate_outputs
20:23:35     "set environment variable GEN_TO_SOURCE or run \"{}\"".format(mismatch, update_cmd))
20:23:35 RuntimeError: Source files: ['TensorMethods.h'] did not match generated files.  To update the source files, set environment variable GEN_TO_SOURCE or run "cp /var/lib/jenkins/workspace/build/aten/src/ATen/core_tmp/TensorMethods.h /var/lib/jenkins/workspace/cmake/../aten/src/ATen/core"
20:23:35 [ 31%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_message.cc.o
20:23:35 gmake[2]: *** [caffe2/CMakeFiles/generate-torch-sources.dir/build.make:208: aten/src/ATen/CPUType.cpp] Error 1
20:23:35 gmake[2]: *** Deleting file 'aten/src/ATen/CPUType.cpp'
```

**Comment by ezyang** [comment]
Well, it's also being reported in the lint builds:

https://dev.azure.com/pytorch/PyTorch/_build/results?buildId=7732&view=logs&jobId=56eb8f65-5d0f-5e57-50ba-12b0255c5704

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -3788,6 +3788,8 @@
   dispatch:
     CPU: legacy::cpu::_th_index_select
     CUDA: legacy::cuda::_th_index_select
+    SparseCPU: index_select_sparse
+    SparseCUDA: index_select_sparse
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: I guess, the only possible explanation is that this has caused some of the generated files to change. Are you sure that you don't get this error locally? A matter of last resort would be to just manually apply the diff yourself, I guess.

@gchanan hurry please, I don't want to have checked in generated files anymore XD

Comment (pearu) [reply aten/src/ATen/native/native_functions.yaml]: That is the thing, locally and in most CI cases there is no error.

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: I figured it out: some of our jobs default to merge with master and that's what is causing the problem.

**Comment by ezyang** [comment]
Should be good now, let's see if CI agrees.

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (pearu) [reply aten/src/ATen/native/native_functions.yaml]: Great, thanks!

File: aten/src/ATen/gen.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/gen.py]: @ezyang @zou3519 Do you want to keep the diff output code for debugging? Or shall I undo these changes to gen.py?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**